discovery_engine_core/providers/src/expression.rs

```diff
@@ -0,0 +1,376 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+//! Expressions
+
+use std::collections::BTreeSet;
+
+use itertools::{intersperse, Either, Itertools};
+use maplit::btreeset;
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub(crate) enum Expr {
+    // contained `Expr` must not be an `And`
+    And(BTreeSet<Expr>),
+    // contained `Expr` must not be an `Or`
+    Or(BTreeSet<Expr>),
+    Value(String),
+}
+
+impl Expr {
```

Comment:
It may be worth also implementing the following, to avoid all those explicit `.clone()` calls :D 

```rust
impl Into<Expr> for &Expr {
    fn into(self) -> Expr {
        self.clone()
    }
}
```

Reply:
Unfortunately does not work, with this we can only remove a `clone` when it is used only on right hand side.
`a.clone().and(b.clone().and(c.clone()))` can simplified only to `a.clone().and(b.clone().and(&c))`.

`and` and `or` are taking `self` so we can reuse the btree without cloning.

Reply:
Mmh, for me the following works fine with that extra `impl`:

```rust
&a.and(&b.or(&c)).build()
```

What am I missing?

Reply:
I'll try it again, but I added it and modified a test and it wasn't working

Reply:
If you have sequences that all start with `&a`, you do need to clone the first one. Maybe that's it?

```rust
&a.clone().and(&b.or(&c)).build()
```

All the inner ones should definitely work, though. At least in my scratchpad :D

Reply:
Did you try to have two?
```rust
&a.and(&b.or(&c)).build()
&a.and(&b.or(&c)).build()
```
on the second I get the error that `a` and `b` were moved on the previous line.

Reply:
The inner has the same problem because rust is implicitly defer the reference to allow `or` to take ownership of `b`, there is no `Into` on the left hand side so it should not even call the new impl

Reply:
Ah, I see the issue. It works in this case:

```rust
#[test]
    fn test_build_and_or() {
        let a = Expr::Value("a".to_string());
        let b = Expr::Value("b".to_string());
        let c = Expr::Value("c".to_string());
        let d = Expr::Value("d".to_string());

        // (a /\ b) \/ c
        assert_eq!(
            "(a AND b) OR c",
            a.clone().and(&b).or(&c).build()
        );

        // a /\ (b \/ c)
        assert_eq!(
            "(b OR c) AND a",
            a.clone().and(&b.or(&c)).build()
        );
    }
```

But not for e.g. `a.clone().or(&b.and(&c))`. Didn't test far enough. Fair enough, in that case one would likely need something more complex. Not worth the effort.